File: 0207-course-schedule/0207-course-schedule.go
func dfs(node int, visited map[int]int, adj_list map[int][]int) bool {
    
    visited[node] = 1
        
    for _,v := range adj_list[node] {
        
        if  visited[v] == 0 {
            res := dfs(v,visited,adj_list)
            if res == false {
            return false
            } 
            
        } else if visited[v] == 1 {
                return false
            }    
        
        
    }
    
    visited[node] = 2
    
    return true    
    
}

func canFinish(numCourses int, prerequisites [][]int) bool {
    adj_list := make(map[int][]int)
    visited := make(map[int]int)
    
    for _, value := range prerequisites {
        u := value[0]
        v := value[1]
        adj_list[v] = append(adj_list[v],u)
    }
    
    for node := range adj_list {
        if visited[node] == 0 {
            res := dfs(node,visited,adj_list)
            if res == false {
                return false
            }
        }
    }
    
    return true
    
}

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def dfs_cycle(self,node,vis,adj):
        vis[node] = 1
        
        for neighbour in adj[node]:
            if(vis[neighbour] == 1):
                return True
            if(vis[neighbour] == 0 and self.dfs_cycle(neighbour,vis,adj)):
                return True
        
        vis[node] = 2
        return False
                
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        adj = [[] for _ in range(numCourses)]
        vis = [0]*numCourses
        for p in prerequisites:
            adj[p[1]].append(p[0])
        
        for i in range(numCourses):
            if(self.dfs_cycle(i,vis,adj)):
                return False
        
        return True
        